classic/docs/example-gallery/python-tutor.md

```diff
@@ -0,0 +1,994 @@
+      ]}
+      opId={"pythonTutorFrame"}
+      objects={[
+        {
+          nextObject: { opId: "o2" },
+          objectType: "tuple",
+          value: "1",
+          opId: "o1",
+        },
+        {
+          nextObject: { opId: "o3" },
+          objectType: "tuple",
+          value: "2",
+          opId: "o2",
+        },
+        { nextObject: null, objectType: "tuple", value: "3", opId: "o3" },
+      ]}
+      rows={[
+        { depth: 0, nodes: ["", "o2", "o3"] },
+        { depth: 1, nodes: ["o1", "", ""] },
+      ]}
+    />
+  </SVG>
+);
+```
+
+## Generalizing Python Tutor
+
+Python Tutor diagrams are very flexible, and the content represented in the diagram can be very diverse. For example, objects could have arbitrary length. Objects themselves could simply store pointers to other primitive types. In the following sections, we will explore how to generalize the Bluefish Python Tutor diagram to flexibility handle these situations.
+
+### Redefining the Components of the Object Type
+
+First, let's modify our tuple Objects. We see that there are two cases we should consider: when the tuple is a pointer to another object, or when the tuple contains a value.
+
+```tsx live noInline
+const ObjectsV2 = forwardRef(function _Objects(
+  { objectType, objectValues, objectId },
+  ref
+) {
+  // objectValues: list with two entries; first entry is type of first box, second entry is type of second box
+
+  const objectTypeRef = useRef(null);
+  const objectRef = useRef(null);
+
+  const boxZeroRef = useRef(null);
+  const boxOneRef = useRef(null);
+  const boxZeroValueRef = useRef(null);
+  const boxOneValueRef = useRef(null);
+  const zeroRef = useRef(null);
+  const oneRef = useRef(null);
+
+  const fontFamily = "verdana, arial, helvetica, sans-serif";
+
+  return (
+    <Group ref={ref} name={objectId}>
+      <Text
+        ref={objectTypeRef}
+        contents={objectType}
+        fontFamily={fontFamily}
+        fontSize={"16px"}
+        fill={"grey"}
+      />
+
+      {/* separate names for each rectangle so that the arrow can go from the center of pointer to the center left of pointed */}
+      <Group ref={objectRef}>
+        <Rect
+          ref={boxOneRef}
+          name={`elm_1_${objectId}`}
+          height={60}
+          width={70}
+          fill={"#ffffc6"}
+          stroke={"grey"}
+        />
+        <Rect
+          ref={boxZeroRef}
+          name={`elm_0_${objectId}`}
+          height={60}
+          width={70}
+          fill={"#ffffc6"}
+          stroke={"grey"}
+        />
+
+        {/* Generate Text in Object based on whether it is a string or a pointer */}
+        {objectValues[0].type == "string" ? (
+          <Text
+            ref={boxZeroValueRef}
+            contents={objectValues[0].value}
+            fontSize={"24px"}
+            fill={"black"}
+          />
+        ) : (
+          <Text ref={boxZeroValueRef} contents={""} fill={"none"} />
```

Comment:
Does something break if e.g. this were a 0-width, 0-height rect? I wonder if we need a `null`/`Fragment` Bluefish component so that Ref resolution still works even though we don't want to render anything...

Reply:
I'm not sure if this is what you mean - but if I change either the object associated with boxOneRef or boxZeroRef, the diagram will still render and there's no error. It just looks a bit weird because the background yellow box is missing, so the text does not have anything to align with.